### PR TITLE
Merge 2.16.x up into 3.0.x

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

**Comment by greg0ire** [comment]
@derrabus what should be done w.r.t https://github.com/doctrine/dbal/pull/5838 on the ORM?

**Comment by greg0ire** [comment]
Ah I think I know: we should replace the `int` in AbstractSQLExecutor with `ArrayParameterType::*` (it's an integer with DBAL 3, and it's an enum with DBAL 4. I hope SA undestands `Enum::*` the same way it understands `Enum` :crossed_fingers:

File: lib/Doctrine/ORM/Query/Exec/AbstractSqlExecutor.php [hunk 1]
```diff
@@ -49,10 +53,9 @@ public function removeQueryCacheProfile(): void
     /**
      * Executes all sql statements.
      *
-     * @param Connection $conn The database connection that is used to execute the queries.
-     * @psalm-param list<mixed>|array<string, mixed> $params The parameters.
-     * @psalm-param array<int, int|string|Type|null>|
-     *              array<string, int|string|Type|null> $types The parameter types.
+     * @param Connection                       $conn   The database connection that is used to execute the queries.
+     * @param list<mixed>|array<string, mixed> $params The parameters.
+     * @psalm-param WrapperParameterTypeArray  $types  The parameter types.
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Query/Exec/AbstractSqlExecutor.php:58]: Ported as https://github.com/doctrine/orm/pull/10796

**Comment by greg0ire** [comment]
@mpdude please review

**Comment by mpdude** [comment]
I don’t know what to specifically watch out for which these merge-ups 🙁

**Comment by mpdude** [comment]
Did you have to resolve any conflicts, or was it a smooth merge?

**Comment by greg0ire** [comment]
It wasn't smooth at all, far from it :sweat: 

I wish Github would indicate where conflict happened, but I think right now the only way is to compare the diff of this PR with https://github.com/doctrine/orm/compare/3.0.x...2.16.x

**Review by derrabus — APPROVED** [review]

**Comment by mpdude** [comment]
🫣

what is the source of the conflicts? Refactorings on 3.0?

**Comment by greg0ire** [comment]
Yes, mostly. Sometimes it is a dumb as cs rules that are not applicable on 3.x, for instance trailing commas after the last argument passed to a method. Some other times, it's more complicated :P

**Pull request merged**